### ISSUE TITLE
Compile time ViewRegistry mechanism for kotlin

### ISSUE THREAD

**Issue by rjrjr** [square/workflow#1205]
Swift was able to drop ViewRegistry in favor of a compile time pattern. Now that compiler plugins are here, shouldn't kotlin be able to do the same?

**Comment by zach-klippenstein** [comment]
On a scale of "fancy lint check" to "productionize keep-87", what exactly are you proposing? Both those options are significantly more effort and maintenance commitment than what swift has.

**Comment by rjrjr** [comment]
I have no real idea, but lint check sounds more like it. Just wanted a
place to track any ideas that occur.

**Comment by zach-klippenstein** [comment]
We've collected quite a few lint check issues. It would be interesting to see if we could build a thin compiler plugin layer that would expose workflow-specific syntax elements, and then implement a lot of checks on top of that.

Or build a similar layer on top of ktlint's API for that matter. I'm not sure which one is better to work with. 

@vRallev Have you lint checks? How would you compare lint or ktlint's API to the compiler plugin API for analysing expressions?

**Comment by vRallev** [comment]
I'm missing context what Swift is doing. Although you can, I wouldn't implement Lint checks as a compiler plugin. KtLint works with Psi elements, like compiler plugins do during the analysis phase. A compiler plugin feels like an overkill. It would run for every build whereas you should run Lint checks only manually.

**Comment by zach-klippenstein** [comment]
Good point. There are some other things we could do with a compiler plugin, but there's no reason we couldn't have both.
